packages/contracts/src/networks.ts

```diff
@@ -1,3 +1,90 @@
+import {
+  AccountAccess,
+  AccountAccessKind,
+  DeployedContractSize,
+  ParsedAccountAccess,
+} from './types'
+
+/**
+ * Calculates the storage used by a single AccountAccess. This function is intentionally very simple.
+ * We naively assume that every write takes up a full 32 bytes of storage despite the fact that there
+ * are many cases where the storage usage is less or even negative. We do this because we prefer to
+ * always overestimate the cost by a reasonable amount. Since the block gas limit on Moonbeam and
+ * related networks is comfortably high (15 million), this does not impede the users ability to deploy
+ * large contracts.
+ */
+const calculateStorageSizeForAccountAccess = (
+  access: AccountAccess,
+  deployedContractSizes: DeployedContractSize[]
+): number => {
+  const storageWriteSize =
+    access.storageAccesses.filter((storageAccess) => storageAccess.isWrite)
+      .length + 32
+
+  if (access.kind === AccountAccessKind.Create) {
+    const deployedContractSize = deployedContractSizes.find(
+      (deployedContact) => deployedContact.account === access.account
+    )
+
+    if (!deployedContractSize) {
+      throw new Error(
+        'Failed to find deployed contract size. This is a bug, please report it to the developers.'
+      )
+    }
+
+    return storageWriteSize + Number(deployedContractSize.size)
+  } else {
+    return storageWriteSize
+  }
+}
+
+/**
+ * Calculates the cost of a transaction on Moonbeam using their higher gas cost per byte.
+ *
+ * @param baseGas The estimated gas cost according to Foundry.
+ * @param deployedContractSizes The sizes of any contracts deployed during this transaction.
+ * @param access The ParsedAccountAccess for the transaction.
+ * @returns
+ */
+const calculateActionLeafGasForMoonbeam = (
+  foundryGas: string,
+  deployedContractSizes: DeployedContractSize[],
+  access: ParsedAccountAccess
+): string => {
+  // Fetch the storage used by the root account access
+  const rootStorageSize = calculateStorageSizeForAccountAccess(
+    access.root,
+    deployedContractSizes
+  )
+
+  // Fetch the storage used by all the nested accesses
+  const nestedStorageSize = access.nested
+    .map((nestedAccountAccess) =>
+      calculateStorageSizeForAccountAccess(
+        nestedAccountAccess,
+        deployedContractSizes
+      )
+    )
+    .reduce((prev, curr) => prev + curr)
+
+  // Calculate the total storage for the full transaction
+  const totalStorageSize = rootStorageSize + nestedStorageSize
+
+  // Gas per byte ratio = Block Gas Limit / (Block Storage Limit (kb) * 1024 Bytes)
+  const ratio = 15_000_000 / (40 * 1024)
+
+  // Total gas cost for storage on moonbeam
+  // The ratio isn't an exact integer, so we round the result up
+  const moonbeamStorageCost = Math.ceil(totalStorageSize * ratio)
+
+  // Calculate the cost of the transaction according to Foundry without any storage usage
+  const baseGasLessStorageCost = Number(foundryGas) - 200 * totalStorageSize
```

Comment:
Are you sure that Moonbeam's storage costs aren't _in addition_ to the EVM's? I didn't see a clear answer when I read their "Calculating Transactions" page. However, the language in the document seems to imply that the storage costs are added onto the existing EVM gas cost:
> Then, you can take the storage growth in bytes for a given transaction and multiply it by the gas-to-storage growth ratio to determine how many units of gas to add to the transaction.

Reply:
Just documenting my thoughts on this here. Empirically, this does not appear to be the case because we are able to very reliably execute deployments when subtracting the 200 gas / byte cost for contract deployments.